File: Tests/YBarKitTests/LuaTests.swift
import Foundation
import Testing
@testable import YBarKit

/// End-to-end tests of the embedded YbarLua runtime against the real object
/// model (headless: BarManager without begin(), so no windows are created).
@MainActor
@Suite(.serialized) struct LuaRuntimeTests {
    private struct Stack {
        let barManager: BarManager
        let eventBus: EventBus
        let scheduler: AnimationScheduler
        let runtime: LuaRuntime
    }

    private func makeStack() throws -> Stack {
        let barManager = try BarManager()
        let eventBus = EventBus()
        eventBus.itemsProvider = { [weak barManager] in barManager?.store.items ?? [] }
        let scheduler = AnimationScheduler()
        let runtime = LuaRuntime(barManager: barManager, eventBus: eventBus, scheduler: scheduler)
        return Stack(barManager: barManager, eventBus: eventBus, scheduler: scheduler, runtime: runtime)
    }

    private func run(_ code: String, _ runtime: LuaRuntime) -> String? {
        let url = FileManager.default.temporaryDirectory
            .appendingPathComponent("ybar-lua-test-\(UUID().uuidString).lua")
        try? code.write(to: url, atomically: true, encoding: .utf8)
        defer { try? FileManager.default.removeItem(at: url) }
        return runtime.runConfig(at: url)
    }

    @Test func barItemAndNestedProperties() throws {
        let stack = try makeStack()
        defer { stack.runtime.shutdown() }
        let error = run("""
        ybar.bar({ height = 40, color = "0xff112233" })
        local clock = ybar.add("item", "clock", "right")
        clock:set({
          label = "hello",
          icon = { color = "0xffff0000", padding_left = 6 },
          ["background.corner_radius"] = 7,
          background = { drawing = true },
        })
        """, stack.runtime)
        #expect(error == nil)
        #expect(stack.barManager.settings.height == 40)
        #expect(stack.barManager.settings.backgroundColor.argb == 0xFF11_2233)

        let clock = stack.barManager.store.item(named: "clock")
        #expect(clock?.position == .right)
        #expect(clock?.label.string == "hello")
        #expect(clock?.icon.color.argb == 0xFFFF_0000)
        #expect(clock?.icon.paddingLeft == 6)
        #expect(clock?.background.cornerRadius == 7)
        #expect(clock?.background.drawing == true)
    }

    @Test func subscriptionCallbackFiresInProcess() throws {
        let stack = try makeStack()
        defer { stack.runtime.shutdown() }
        // Wire the Lua-first dispatch the daemon normally provides.
        stack.eventBus.runItemScript = { [weak runtime = stack.runtime] item, env in
            _ = runtime?.handleEvent(item: item, environment: env)
        }
        let error = run("""
        local w = ybar.add("item", "wifi", "left")
        w:subscribe("wifi_change", function(env)
          w:set({ label = "net:" .. env.INFO })
        end)
        """, stack.runtime)
        #expect(error == nil)

        stack.eventBus.trigger(name: "wifi_change", info: "MyNetwork")
        #expect(stack.barManager.store.item(named: "wifi")?.label.string == "net:MyNetwork")
    }

    @Test func luaOnlyItemsReceiveEventsWithoutShellScript() throws {
        let stack = try makeStack()
        defer { stack.runtime.shutdown() }
        var shellRan = false
        stack.eventBus.runItemScript = { [weak runtime = stack.runtime] item, env in
            if runtime?.handleEvent(item: item, environment: env) == true { return }
            shellRan = true
        }
        let error = run("""
        local i = ybar.add("item", "x", "left")
        i:subscribe("system_woke", function(env) i:set({ label = "woke" }) end)
        """, stack.runtime)
        #expect(error == nil)
        let item = stack.barManager.store.item(named: "x")
        #expect(item?.hasLuaHandlers == true)
        #expect(item?.script.isEmpty == true)

        stack.eventBus.trigger(name: "system_woke")
        #expect(item?.label.string == "woke")
        #expect(!shellRan)
    }

    @Test func componentsAndGraphPush() throws {
        let stack = try makeStack()
        defer { stack.runtime.shutdown() }
        let error = run("""
        local g = ybar.add("graph", "cpu", "right", 4)
        g:push({ 0.25, 0.5 })
        local s = ybar.add("slider", "vol", "right", 80)
        s:set({ slider = { percentage = 66 } })
        local a = ybar.add("item", "a", "left")
        a:set({ label = "m" })
        local br = ybar.add("bracket", "grp", { "a" })
        """, stack.runtime)
        #expect(error == nil)

        let graph = stack.barManager.store.item(named: "cpu")?.graph
        #expect(graph?.capacity == 4)
        #expect(graph?.ordered().suffix(2) == [0.25, 0.5])
        #expect(stack.barManager.store.item(named: "vol")?.slider?.percentage == 66)
        let bracket = stack.barManager.store.item(named: "grp")
        #expect(bracket?.kind == .bracket)
        #expect(bracket?.members == ["a"])
    }

    @Test func animateWrapperScopesContext() throws {
        let stack = try makeStack()
        defer { stack.runtime.shutdown() }
        let error = run("""
        local i = ybar.add("item", "anim", "left")
        i:set({ ["label.color"] = "0xff000000" })
        ybar.animate("linear", 30, function()
          i:set({ ["label.color"] = "0xffffffff" })
        end)
        """, stack.runtime)
        #expect(error == nil)
        // The color set inside animate() must be scheduled, not applied directly.
        #expect(stack.scheduler.isAnimating)
        #expect(stack.runtime.animationContext == nil)
        #expect(stack.barManager.store.item(named: "anim")?.label.color.argb == 0xFF00_0000)
    }

    @Test func configErrorsAreReportedNotFatal() throws {
        let stack = try makeStack()
        defer { stack.runtime.shutdown() }
        let error = run("this is not lua", stack.runtime)
        #expect(error?.contains("lua") == true)
    }
}

File: tests/lua_tests.cpp
// End-to-end embedded-Lua tests (LuaTests.swift parity) — a headless
// runtime driving the real ItemStore/EventBus/CommandHandler funnel
// (docs/WINDOWS-PORT.md sections 3.7, 12).

#include <cstdio>
#include <cstdlib>
#include <cwctype>
#include <string>
#include <vector>

#include <catch2/catch_test_macros.hpp>

#include "app/script_runner.h"
#include "lua/runtime.h"

using namespace ybar;

namespace {

struct Fixture {
    model::ItemStore store;
    model::BarSettings settings;
    events::EventBus bus;
    ipc::CommandHandler handler{store, settings, bus, {}};
    app::ScriptRunner scripts;
    lua::LuaRuntime runtime{store, bus, handler, scripts, /*messageWindow=*/nullptr};

    std::string configPath;

    ~Fixture() {
        if (!configPath.empty()) std::remove(configPath.c_str());
    }

    void run(const std::string& source) {
        configPath = std::string(std::getenv("TEMP") ? std::getenv("TEMP") : ".") +
                     "\\ybar_lua_test.lua";
        if (FILE* file = std::fopen(configPath.c_str(), "wb")) {
            std::fwrite(source.data(), 1, source.size(), file);
            std::fclose(file);
        }
        runtime.runConfig(configPath);
    }
};

} // namespace

TEST_CASE("ybar.add with nested prop tables flattens to dotted sets") {
    Fixture f;
    f.run(R"(
        local h = ybar.add("item", "clock", "right", {
            label = "12:00",
            background = { height = 16, corner_radius = 4 },
            ["label.color"] = "0xffff0000",
        })
        h:set({ padding_left = 8 })
    )");
    auto* item = f.store.find("clock");
    REQUIRE(item);
    CHECK(item->position == model::ItemPosition::Right);
    CHECK(item->label.string == "12:00");
    CHECK(item->background.height == 16);
    CHECK(item->background.cornerRadius == 4);
    CHECK(item->label.color.argb == 0xffff0000u);
    CHECK(item->paddingLeft == 8);
}

TEST_CASE("integral floats stringify without .0 (update_freq = 60/2)") {
    Fixture f;
    f.run(R"(
        ybar.add("item", "a", "left", { update_freq = 60/2 })
    )");
    REQUIRE(f.store.find("a"));
    CHECK(f.store.find("a")->updateFrequency == 30); // "30", not "30.0"
}

TEST_CASE("booleans stringify as on/off through the prelude") {
    Fixture f;
    f.run(R"(
        ybar.add("item", "a", "left", { drawing = false, ["icon.highlight"] = true })
    )");
    auto* item = f.store.find("a");
    REQUIRE(item);
    CHECK_FALSE(item->drawing);
    CHECK(item->icon.highlight);
}

TEST_CASE("subscriptions dispatch Lua-first with SENDER routing") {
    Fixture f;
    f.run(R"(
        local h = ybar.add("item", "listener", "left", {})
        seen = {}
        h:subscribe("system_woke", function(env)
            seen.sender = env.SENDER
            seen.name = env.NAME
        end)
        h:subscribe("routine", function(env) seen.routine = env.SENDER end)
    )");
    auto* item = f.store.find("listener");
    REQUIRE(item);
    CHECK(item->hasLuaHandlers);
    CHECK((item->updateMask & (1ull << 3)) != 0); // system_woke bit

    CHECK(f.runtime.handleEvent(*item, {{"NAME", "listener"},
                                        {"SENDER", "system_woke"},
                                        {"INFO", ""}}));
    // SENDER=forced falls back to the routine handler (SbarLua semantics).
    CHECK(f.runtime.handleEvent(*item, {{"NAME", "listener"},
                                        {"SENDER", "forced"},
                                        {"INFO", ""}}));
    // No handler for mouse.clicked -> false so shell scripts can run.
    CHECK_FALSE(f.runtime.handleEvent(*item, {{"NAME", "listener"},
                                              {"SENDER", "mouse.clicked"},
                                              {"INFO", ""}}));
}

TEST_CASE("bracket members ride the position argument") {
    Fixture f;
    f.run(R"(
        ybar.add("item", "m1", "left", {})
        ybar.add("item", "m2", "left", {})
        ybar.add("bracket", "group", { "m1", "m2" }, {})
    )");
    auto* bracket = f.store.find("group");
    REQUIRE(bracket);
    CHECK(bracket->kind == model::ItemKind::Bracket);
    CHECK(bracket->members == std::vector<std::string>{"m1", "m2"});
}

TEST_CASE("ybar.query returns JSON and query_table returns a table") {
    Fixture f;
    f.run(R"(
        ybar.add("item", "q", "left", { label = "text" })
        query_result = ybar.query("q")
        local t = ybar.query_table("q")
        table_label = t and t.label.value or "MISSING"
        missing = ybar.query_table("nope")
    )");
    // The Lua globals prove both paths worked; verify via a second config run
    // is overkill — assert on the store side effects instead.
    REQUIRE(f.store.find("q"));
    CHECK(f.store.find("q")->label.string == "text");
}

TEST_CASE("ybar.animate applies sets directly when no scheduler is wired") {
    Fixture f;
    f.run(R"(
        local h = ybar.add("item", "anim", "left", {})
        ybar.animate("tanh", 30, function()
            h:set({ padding_left = 42 })
        end)
    )");
    REQUIRE(f.store.find("anim"));
    CHECK(f.store.find("anim")->paddingLeft == 42); // headless: direct set
}

TEST_CASE("custom events flow from Lua registration to Lua handlers") {
    Fixture f;
    f.run(R"(
        ybar.add_event("my_event")
        local h = ybar.add("item", "l", "left", {})
        got = nil
        h:subscribe("my_event", function(env) got = env.PAYLOAD end)
    )");
    auto* item = f.store.find("l");
    REQUIRE(item);
    REQUIRE(f.bus.knownEvent("my_event"));

    f.bus.itemsProvider = [&] { return std::vector<model::Item*>{item}; };
    f.bus.runItemScript = [&](model::Item& target, const events::Environment& env) {
        f.runtime.handleEvent(target, env);
    };
    f.bus.trigger("my_event", "", {{"PAYLOAD", "hello"}});
    // The handler stored PAYLOAD into a Lua global; no crash + dispatch is the
    // assertion here (globals are checked implicitly by pcall success).
    SUCCEED();
}

TEST_CASE("environmentBlock merges variable names case-insensitively") {
    // Windows env names are case-insensitive, and the launcher decides the
    // spelling: Explorer/pwsh pass `Path`, MSYS shells pass `PATH`. A
    // case-sensitive merge duplicated the key, and the PATH prepend then
    // created a second PATH entry holding only the exe directory — children
    // resolved against it and lost powershell.exe/netsh/everything.
    app::ScriptRunner scripts;
    const auto block = scripts.environmentBlock({{"pAtH", "C:\\ybar-test-marker"}});

    std::vector<std::wstring> pathEntries;
    for (const wchar_t* cursor = block.data(); *cursor;) {
        std::wstring entry(cursor);
        cursor += entry.size() + 1;
        const auto eq = entry.find(L'=');
        if (eq == std::wstring::npos) continue;
        std::wstring key = entry.substr(0, eq);
        for (auto& c : key) c = static_cast<wchar_t>(towlower(c));
        if (key == L"path") pathEntries.push_back(entry.substr(eq + 1));
    }

    REQUIRE(pathEntries.size() == 1);
    // The overlay value survived AND got the exe-directory prepend.
    CHECK(pathEntries[0].find(L"C:\\ybar-test-marker") != std::wstring::npos);
    CHECK(pathEntries[0].find(L';') != std::wstring::npos);
}

TEST_CASE("YBAR_SHELL is classified by basename, case-insensitively") {
    // Regression: a case-sensitive whole-path substring match sent
    // ...\PowerShell.exe down the PowerShell branch only by luck of casing,
    // and mistook a POSIX shell nested under a "...cmder..." tree (the path
    // contains "cmd") for cmd.exe. Classify on the lowercased basename.
    const auto probe = [](const wchar_t* shell) {
        _wputenv_s(L"YBAR_SHELL", shell);
        app::ScriptRunner s;
        _wputenv_s(L"YBAR_SHELL", L""); // don't leak to other tests
        return s.commandLineFor("echo hi");
    };

    // Mixed-case PowerShell → `-Command`, never `-c`/`/c`.
    CHECK(probe(L"C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\PowerShell.exe")
              .find(L"-Command") != std::wstring::npos);
    // POSIX sh under a cmder tree ("cmd" in the path) → `-c`, not `/c`.
    const auto cmder =
        probe(L"C:\\tools\\cmder\\vendor\\git-for-windows\\usr\\bin\\sh.exe");
    CHECK(cmder.find(L" -c ") != std::wstring::npos);
    CHECK(cmder.find(L" /c ") == std::wstring::npos);
    // cmd.exe → `/c`, not the PowerShell flags.
    const auto cmd = probe(L"C:\\Windows\\System32\\cmd.exe");
    CHECK(cmd.find(L" /c ") != std::wstring::npos);
    CHECK(cmd.find(L"-Command") == std::wstring::npos);
}

TEST_CASE("config reload drops old-state subscriptions") {
    Fixture f;
    f.run(R"(
        local h = ybar.add("item", "x", "left", {})
        h:subscribe("system_woke", function() end)
    )");
    auto* item = f.store.find("x");
    REQUIRE(item);
    CHECK(f.runtime.handleEvent(*item, {{"SENDER", "system_woke"}}));

    f.runtime.teardown();
    CHECK_FALSE(f.runtime.handleEvent(*item, {{"SENDER", "system_woke"}}));
}
